cpp/src/arrow/python/arrow_to_python.cc

```diff
@@ -165,20 +165,19 @@ Status GetValue(PyObject* context, const Array& arr, int64_t index, int32_t type
 
 #define DESERIALIZE_SEQUENCE(CREATE_FN, SET_ITEM_FN)                                  \
   const auto& data = static_cast<const UnionArray&>(array);                           \
-  int64_t size = array.length();                                                      \
   ScopedRef result(CREATE_FN(stop_idx - start_idx));                                  \
-  auto types = std::make_shared<Int8Array>(size, data.type_ids());                    \
-  auto offsets = std::make_shared<Int32Array>(size, data.value_offsets());            \
+  const uint8_t* type_ids = data.raw_type_ids();                                      \
+  const int32_t* value_offsets = data.raw_value_offsets();                            \
   for (int64_t i = start_idx; i < stop_idx; ++i) {                                    \
     if (data.IsNull(i)) {                                                             \
       Py_INCREF(Py_None);                                                             \
       SET_ITEM_FN(result.get(), i - start_idx, Py_None);                              \
     } else {                                                                          \
-      int64_t offset = offsets->Value(i);                                             \
-      int8_t type = types->Value(i);                                                  \
+      int64_t offset = value_offsets[i];                                              \
+      uint8_t type = type_ids[i];                                                     \
       PyObject* value;                                                                \
       RETURN_NOT_OK(                                                                  \
-          GetValue(context, *data.child(type), offset, type, base, tensors, &value)); \
+          GetValue(context, *data.UnsafeChild(type), offset, type, base, tensors, &value)); \
```

Comment:
I'm surprised this alone makes such an impact. How does this compare with `*data.child(type).get()`?

Reply:
This is called a lot of times and each time it is called, the reference counter needs to be incremented and decremented. These are the times to deserialize a list of200000 ints for 1000 times:

With just child: 15.2s
With UnsafeChild: 11.5s
With child.get: 14.8s (same as just child up to measurement error)
Inlining child: 14.9s (same as above)

Seems like the compiler just can't optimize the reference counter operations out.

Reply:
Thanks, I missed the detail about the shared_ptr being copied by `child`. Makes sense, thanks